clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -70,6 +75,17 @@ public AbstractConfig(ConfigDef definition, Map<?, ?> originals) {
         this(definition, originals, true);
     }
 
+    /**
+     * Called directly after user configs got parsed (and thus default values got set).
+     * This allows to change default values for "secondary defaults" if required.
+     *
+     * @param parsedValues unmodifiable map of current configuration
+     * @return a map of updates that should be applied to the configuration (will be validated to prevent bad updates)
+     */
+    protected Map<String, Object> postProcessParsedConfig(Map<String, Object> parsedValues) {
```

Comment:
@guozhangwang the original solution meant that if you iterated over the configs, you'd not get the right default and the logged config values would also be wrong. That's why I suggested doing it this way.